File: problem-001/initial_solution/efficient_iteration.py
"""
Author: Kona Fatsea
Date: 07/02/2025

Efficient Iteration (Inclusion - Exclusion) solution for problem-001 of Euler Project.

Problem Statement:
Find the sum of all multiples of 3 and 5 below a given number n.

The solutions include:
1) Brute Force Iteration
2) Efficient Iteration (Inclusion-Exclusion Set Theory) <------
3) Arithmetic Sequence
"""


def sum_multiples(n):
    """
    Calculate the sum of multiples of 3 and 5 below n.

    Args:
    n (int): Upper limit (exclusive) for considering multiples.

    Returns:
    int: Sum of all multiples of 3 or 5 below n.
    """
    iterations = 0  # To compare iterations of different methods
    total_sum = 0
    for i in range(0, n, 3):
        total_sum += i
        iterations += 1
    for i in range(0, n, 5):
        total_sum += i
        iterations += 1
    for i in range(0, n, 15):
        total_sum -= i
        iterations += 1
    print(f"Number of Iterations: {iterations}")
    return total_sum


def main():
    """
    Main function to handle user input and calculate the sum of multiples

    Returns:
    tuple: A tuple containing the result and input number.
    """
    n = int(input("Enter an integer: "))
    result = sum_multiples(n)
    print(f"The sum of multiples of 3 or 5 below {n} is: {result}")
    return result, n


if __name__ == "__main__":
    sum_of_multiples, upper_bound = main()

File: problem-001/initial_solution/java/EfficientIteration.java
/* 
Author: Kona Fatsea
Date: 07/02/2025

Solve problem-001 of Euler Project to find the sum of all multiples of 3 and 5
below a given number n using efficient iteration in Java.
*/

import java.util.Scanner;                         



public class EfficientIteration { // Class declaration: the class name should start with an uppercase letter by convention.
    
    public static void main(String[] args) { // Main method - Entry point of the program
        Scanner scanner = new Scanner(System.in); // Initialize the Scanner object to read input from the console
        
        System.out.print("Enter a positive integer: "); // Prompt for input
        int n = scanner.nextInt(); // Read the input number       
        
        // Check if the input is a positive integer
        if (n<=0) {
            System.out.println("Please enter a positive integer.");
        } else {
            // Call sumMultiples and store the result
            int result = sumMultiples(n);
            System.out.printf("The sum of multiples of 3 or 5 below %d is: %d\n", n, result); // Output the result
        }
        
        scanner.close(); // Close the scanner to prevent resource leaks
    }

    /**
     * Calculates the sum of all multiples of 3 or 5 below given n.
     * This method uses straightforward iteration, separately summing multiples of 
     * 3 and 5, ensuring multiples of 15 are not double counted. 
     * 
     * @param n The upper limit (exclusive)
     * @return The sum of all multiples of 3 and 5 below n
     */
    static int sumMultiples(int n){
        int iterations = 0;
        int totalSum = 0;

        // Accumulate multiples of 3
        for (int i = 3; i < n; i+= 3){
            totalSum += i;
            iterations++;
        }
        // Accumulate multiples of 5 
        for (int i = 5; i < n; i += 5){
            if (i % 3 != 0){ // Ensure no double-counting
                totalSum += i;
            }
            iterations++;
        }
       
        System.out.println("Number of Iterations: " + iterations);
        return totalSum; 

    } 
}
 

